File: src/Configuration.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;

internal sealed class Configuration(
    string home,
    string executableDirectory,
    string workingDirectory
)
{
    private readonly HashSet<string> _loaded = new(StringComparer.Ordinal);
    public List<string> Files { get; } = [];

    public static string Expand(string path) =>
        Expand(path, Environment.GetFolderPath(Environment.SpecialFolder.UserProfile));

    public static string Expand(string path, string home)
    {
        if (path == "~")
            path = home;
        else if (
            path.StartsWith("~/", StringComparison.Ordinal)
            || path.StartsWith("~\\", StringComparison.Ordinal)
        )
            path = Path.Combine(home, path[2..]);
        path = Regex.Replace(
            path,
            @"\$(?:\{(\w+)\}|(\w+))|%(\w+)%",
            m =>
            {
                var name = m.Groups.Cast<Group>().Skip(1).First(g => g.Success).Value;
                return Environment.GetEnvironmentVariable(name) ?? m.Value;
            }
        );
        return path;
    }

    public string[] Load(string[] commandLine)
    {
        var layers = new List<List<string>> { LoadLayer(commandLine, workingDirectory) };
        bool Ignore() =>
            layers.SelectMany(x => x).Any(x => x is "--ignore-config" or "--no-config");
        void Add(string? path)
        {
            if (path is null || !File.Exists(path))
                return;
            var args = LoadFile(path);
            if (args is not null)
                layers.Add(args);
        }

        if (!Ignore())
            Add(Path.Combine(executableDirectory, "yt-dlp.conf"));
        if (!Ignore())
        {
            var paths = Options
                .Parse(layers.AsEnumerable().Reverse().SelectMany(x => x), validate: false)
                .Paths;
            var directory = paths.GetValueOrDefault("home", workingDirectory);
            Add(Path.Combine(Expand(directory, home), "yt-dlp.conf"));
        }
        List<string>? userLayer = null;
        if (!Ignore())
        {
            var directories = new List<string>
            {
                Path.Combine(
                    Environment.GetEnvironmentVariable("XDG_CONFIG_HOME")
                        ?? Path.Combine(home, ".config"),
                    "yt-dlp"
                ),
            };
            var appdata =
                Environment.GetEnvironmentVariable("APPDATA")
                ?? Environment.GetEnvironmentVariable("appdata");
            if (!string.IsNullOrEmpty(appdata))
                directories.Add(Path.Combine(appdata, "yt-dlp"));
            directories.Add(Path.Combine(home, ".yt-dlp"));
            var before = layers.Count;
            Add(directories.SelectMany(Candidates).FirstOrDefault(File.Exists));
            if (layers.Count > before)
                userLayer = layers[^1];
        }
        if (!Ignore())
        {
            Add(Candidates("/etc/yt-dlp").FirstOrDefault(File.Exists));
            if (Ignore() && userLayer is not null)
                layers.Remove(userLayer);
        }
        return layers.AsEnumerable().Reverse().SelectMany(x => x).ToArray();
    }

    private static IEnumerable<string> Candidates(string directory)
    {
        var parent = Path.GetDirectoryName(directory)!;
        yield return Path.Combine(parent, "yt-dlp.conf");
        if (Path.GetFileName(directory) == ".yt-dlp")
            yield return Path.Combine(parent, "yt-dlp.conf.txt");
        yield return Path.Combine(directory, "config");
        yield return Path.Combine(directory, "config.txt");
    }

    private List<string>? LoadFile(string path)
    {
        path = Path.GetFullPath(path);
        if (!_loaded.Add(path))
            return null;
        Files.Add(path);
        return LoadLayer(Tokenize(File.ReadAllText(path)), Path.GetDirectoryName(path)!);
    }

    private List<string> LoadLayer(IEnumerable<string> arguments, string directory)
    {
        var own = arguments.ToList();
        var includes = new List<string>();
        for (var i = 0; i < own.Count; i++)
        {
            var arg = own[i];
            if (arg == "--no-config-locations")
                includes.Clear();
            else if (arg.StartsWith("--config-locations=", StringComparison.Ordinal))
                includes.Add(arg.Split('=', 2)[1]);
            else if (arg is "--config-locations" or "--config-location")
            {
                if (++i == own.Count)
                    throw new ArgumentException($"Missing value for {arg}.");
                includes.Add(own[i]);
            }
        }
        var children = new List<List<string>>();
        foreach (var include in includes)
        {
            if (include == "-")
            {
                if (_loaded.Add("<stdin>"))
                    children.Add(LoadLayer(Tokenize(Console.In.ReadToEnd()), directory));
                continue;
            }
            var path = Path.GetFullPath(Expand(include, home), directory);
            if (Directory.Exists(path))
                path = Path.Combine(path, "yt-dlp.conf");
            if (!File.Exists(path))
                throw new ArgumentException($"Config location does not exist: {path}");
            if (LoadFile(path) is { } child)
                children.Add(child);
        }
        return children.AsEnumerable().Reverse().SelectMany(x => x).Concat(own).ToList();
    }

    public static string[] Tokenize(string text)
    {
        var words = new List<string>();
        var word = new StringBuilder();
        char quote = '\0';
        var started = false;
        for (var i = 0; i < text.Length; i++)
        {
            var c = text[i];
            if (quote == '\0' && c == '#')
            {
                while (i < text.Length && text[i] != '\n')
                    i++;
                if (started)
                {
                    words.Add(word.ToString());
                    word.Clear();
                    started = false;
                }
                continue;
            }
            if (c == '\\' && quote != '\'')
            {
                if (++i == text.Length)
                    throw new ArgumentException("Trailing escape in configuration.");
                c = text[i];
                if (quote == '"' && c is not '"' and not '\\')
                    word.Append('\\');
                word.Append(c);
                started = true;
            }
            else if (quote != '\0')
            {
                if (c == quote)
                    quote = '\0';
                else
                    word.Append(c);
            }
            else if (c is '\'' or '"')
            {
                quote = c;
                started = true;
            }
            else if (char.IsWhiteSpace(c))
            {
                if (started)
                {
                    words.Add(word.ToString());
                    word.Clear();
                    started = false;
                }
            }
            else
            {
                word.Append(c);
                started = true;
            }
        }
        if (quote != '\0')
            throw new ArgumentException("Unclosed quote in configuration.");
        if (started)
            words.Add(word.ToString());
        return words.ToArray();
    }
}

File: bun/src/configuration.ts
import { stat } from "node:fs/promises";
import { homedir } from "node:os";
import { basename, dirname, join, resolve } from "node:path";
import { expandPath, parseOptions } from "./options.ts";

export function tokenize(text: string): string[] {
  const words: string[] = [];
  let word = "", quote = "", started = false;
  for (let i = 0; i < text.length; i++) {
    let c = text[i];
    if (!quote && c === "#") {
      while (i < text.length && text[i] !== "\n") i++;
      if (started) { words.push(word); word = ""; started = false; }
      continue;
    }
    if (c === "\\" && quote !== "'") {
      if (++i === text.length) throw new Error("Trailing escape in configuration.");
      c = text[i];
      if (quote === '"' && c !== '"' && c !== "\\") word += "\\";
      word += c; started = true;
    } else if (quote) {
      if (c === quote) quote = ""; else word += c;
    } else if (c === "'" || c === '"') { quote = c; started = true; }
    else if (/\s/.test(c)) {
      if (started) { words.push(word); word = ""; started = false; }
    } else { word += c; started = true; }
  }
  if (quote) throw new Error("Unclosed quote in configuration.");
  if (started) words.push(word);
  return words;
}

export async function readText(path: string): Promise<string> {
  const bytes = await Bun.file(path).bytes();
  if (bytes[0] === 0xff && bytes[1] === 0xfe) return new TextDecoder("utf-16le").decode(bytes);
  if (bytes[0] === 0xfe && bytes[1] === 0xff) return new TextDecoder("utf-16be").decode(bytes);
  return new TextDecoder().decode(bytes);
}

export async function loadConfiguration(commandLine: string[], {
  home = homedir(), executableDirectory = dirname(process.execPath),
  workingDirectory = process.cwd(), systemDirectory = "/etc/yt-dlp", env = process.env,
  stdin = () => Bun.stdin.text(),
} = {}) {
  const loaded = new Set<string>();
  const files: string[] = [];
  async function loadFile(path: string): Promise<string[] | null> {
    path = resolve(path);
    if (loaded.has(path)) return null;
    loaded.add(path); files.push(path);
    return loadLayer(tokenize(await readText(path)), dirname(path));
  }
  async function loadLayer(own: string[], directory: string): Promise<string[]> {
    let includes: string[] = [];
    for (let i = 0; i < own.length; i++) {
      const arg = own[i];
      if (arg === "--no-config-locations") includes = [];
      else if (arg.startsWith("--config-locations=")) includes.push(arg.slice(arg.indexOf("=") + 1));
      else if (arg === "--config-locations" || arg === "--config-location") {
        if (++i === own.length) throw new Error(`Missing value for ${arg}.`);
        includes.push(own[i]);
      }
    }
    const children: string[][] = [];
    for (const include of includes) {
      if (include === "-") {
        if (!loaded.has("<stdin>")) {
          loaded.add("<stdin>"); children.push(await loadLayer(tokenize(await stdin()), directory));
        }
        continue;
      }
      let path = resolve(directory, expandPath(include, home, env));
      if ((await stat(path).catch(() => null))?.isDirectory()) path = join(path, "yt-dlp.conf");
      if (!await Bun.file(path).exists()) throw new Error(`Config location does not exist: ${path}`);
      const child = await loadFile(path);
      if (child) children.push(child);
    }
    return [...children.reverse().flat(), ...own];
  }
  const layers = [await loadLayer(commandLine, workingDirectory)];
  function ignore() { return layers.flat().some(x => x === "--ignore-config" || x === "--no-config"); }
  async function add(path: string | undefined) {
    if (!path || !await Bun.file(path).exists()) return;
    const args = await loadFile(path);
    if (args) layers.push(args);
  }
  async function firstExisting(paths: string[]) {
    for (const path of paths) if (await Bun.file(path).exists()) return path;
  }
  if (!ignore()) await add(join(executableDirectory, "yt-dlp.conf"));
  if (!ignore()) {
    const paths = parseOptions(layers.toReversed().flat(), false).paths;
    await add(join(expandPath(paths.get("home") ?? workingDirectory, home, env), "yt-dlp.conf"));
  }
  let userLayer: string[] | undefined;
  if (!ignore()) {
    const directories = [join(env.XDG_CONFIG_HOME ?? join(home, ".config"), "yt-dlp")];
    const appdata = env.APPDATA ?? env.appdata;
    if (appdata) directories.push(join(appdata, "yt-dlp"));
    directories.push(join(home, ".yt-dlp"));
    const before = layers.length;
    await add(await firstExisting(directories.flatMap(candidates)));
    if (layers.length > before) userLayer = layers.at(-1);
  }
  if (!ignore()) {
    await add(await firstExisting(candidates(systemDirectory)));
    if (ignore() && userLayer) layers.splice(layers.indexOf(userLayer), 1);
  }
  return { args: layers.reverse().flat(), files };
}

function candidates(directory: string): string[] {
  return [join(dirname(directory), "yt-dlp.conf"),
    ...(basename(directory) === ".yt-dlp" ? [join(dirname(directory), "yt-dlp.conf.txt")] : []),
    join(directory, "config"), join(directory, "config.txt")];
}
